File: algorithms/python/779.py
class Solution:
    def kthGrammar(self, N, K):
        """
        :type N: int
        :type K: int
        :rtype: int
        """
        N, K = N-1, K-1
        def helper(N, K):
            if N == 0: return 0
            prev = helper(N-1, K//2)
            return (K+prev)%2
        return helper(N, K)

File: algorithms/rust/779.rs
impl Solution {
    #[allow(dead_code)]
    pub fn kth_grammar(n: i32, k: i32) -> i32 {
        let mut paths = vec![k];
        for _ in 1..n {
            let last = paths.last().unwrap();
            paths.push((last + 1) / 2);
        }
        let mut rslt = 0;
        for path in paths.iter().rev() {
            if path % 2 == 0 {
                rslt = 1 - rslt;
            }
        }
        rslt
    }
}

struct Solution;

fn main() {
    println!("Hello, world!");
}
